Changelog for MLflow 1.13.0 (#3890)
mlflow/mlflow: CHANGELOG.rst

- Add support for ``statsmodels`` flavor (#3304, @olbapjose)
- Add support for nested-run in mlflow R client (#3765, @yitao-li)
- Deploying a model using ``mlflow.azureml.deploy`` now integrates better with the AzureML tracking/registry. (#3419, @trangevi)

Bug fixes and documentation updates:

- When running an MLflow Project on Databricks, the version of MLflow installed on the Databricks cluster will now match the version used to run the Project (#3880, @FlorisHoogenboom)
- Fix bug where metrics are not logged for single-epoch ``tf.keras`` training sessions (#3853, @dbczumar)
- Reject boolean types when logging MLflow metrics (#3822, @HCoban)
- Fix alignment of Keras / ``tf.Keras`` metric history entries when ``initial_epoch`` is different from zero. (#3575, @garciparedes)
- Fix bugs in autologging integrations for newer versions of TensorFlow and Keras (#3735, @dbczumar)
- Drop global ``filterwwarnings`` module at import time (#3621, @jogo)
- Fix bug that caused preexisting Python loggers to be disabled when using MLflow with the SQLAlchemyStore (#3653, @arthury1n)
- Fix ``h5py`` library incompatibility for exported Keras models (#3667, @tomasatdatabricks)

> Small changes, bug fixes and doc updates (#3887, #3882, #3845, #3861, #3855, #3854, #3833, #3828, #3826, #3825, #3800, #3809, #3786, #3794, #3731, #3776, #3760, #3771, #3754, #3750, #3749, #3747, #3736, #3701, #3699, #3658, #3675, @harupy; #3723, @mohamad-arabi; #3650, #3655, @shrinath-suresh; #3850, #3753, @dmatrix; #3664, #3867 @smurching; #3681, @sueann; #3619, @andrewnitu; #3837, @javierluraschi; #3721, @szczeles; #3653, @arthury1n; #3883, #3874, #3870, #3877, #3878, #3815, #3859, #3844, #3703, @dbczumar; #3768, @wentinghu; #3784, @HCoban; #3643, #3649, @arjundc-db)

Reviewer: @harupy Are we sure that the changelog generation script captured all of the relevant PRs? For example, @mohamad-arabi contributed #3860 and #3838, but I don't see either of those reflected here.

Reviewer 2: It looks like the changelog generation script ignores PRs without `rn/xxx` label.

Changelog
=========
1.13 (2020-12-22)
-----------------
MLflow 1.13 includes several major features and improvements:

Features:

New fluent APIs for logging in-memory objects as artifacts:

- Add ``mlflow.log_text`` which logs text as an artifact (#3678, @harupy)
- Add ``mlflow.log_dict`` which logs a dictionary as an artifact (#3685, @harupy)
- Add ``mlflow.log_figure`` which logs a figure object as an artifact (#3707, @harupy)
- Add ``mlflow.log_image`` which logs an image object as an artifact (#3728, @harupy)

UI updates / fixes (#3867, @smurching):

- Add model version link in compact experiment table view
- Add logged/registered model links in experiment runs page view
- Enhance artifact viewer for MLflow models
- Model registry UI settings are now persisted across browser sessions
- Add model version ``description`` field to model version table

Autologging enhancements:

> - Introduce utilities for autologging error tolerance / safety (#3815, dbczumar)

Reviewer 2: will replace #3815 with #3862

> - Introduce utilities for autologging error tolerance / safety (#3682, dbczumar)

Reviewer: The way this is described isn’t really a user-facing change. Can we say something to the effect of “improved robustness of autologging integrations to exceptions” ?

Reviewer 2: Can we simply say `Improve robustness of autologging integrations to exceptions`?

Reviewer: Yep!

> - Add ``exclusive`` configuration option for autologging (#3869, @dbczumar)

Reviewer: Can we add the disable flag to this entry too and link the associated PR?

Reviewer: @harupy I think this should be `disable` and should reference a PR other than #3869 . Right now, this is a duplicate entry

- Set tags on autologged runs for easy identification (and add tags to start_run) (#3847, @dbczumar)
- Add ``log_models`` configuration for autologging (#3663, @mohamad-arabi)

More features and improvements:

> - Retry SQLAlchemy engine creation (#3784, @HCoban)

Reviewer: On second thought, this sqlalchemy one seems like more of a small fix. Let’s move it.

Reviewer 2: Done!

- Allow Keras models to be saved with ``SavedModel`` format (#3552, @skylarbpayne)
- Add support for ``statsmodels`` flavor (#3304, @olbapjose)
- Add support for nested-run in mlflow R client (#3765, @yitao-li)

> - Java client changes to incorporate PreSigned URL support (#3649, @arjundc-db)

Reviewer: This presented url work is databricks-specific and should be moved into the small changes section

Reviewer 2: Done!

- Deploying a model using ``mlflow.azureml.deploy`` now integrates better with the AzureML tracking/registry. (#3419, @trangevi)

Bug fixes and documentation updates:

- When running an MLflow Project on Databricks, the version of MLflow installed on the Databricks cluster will now match the version used to run the Project (#3880, @FlorisHoogenboom)
- Fix bug where metrics are not logged for single-epoch ``tf.keras`` training sessions (#3853, @dbczumar)
- Reject boolean types when logging MLflow metrics (#3822, @HCoban)
- Fix alignment of Keras / ``tf.Keras`` metric history entries when ``initial_epoch`` is different from zero. (#3575, @garciparedes)
- Fix bugs in autologging integrations for newer versions of TensorFlow and Keras (#3735, @dbczumar)
- Drop global ``filterwwarnings`` module at import time (#3621, @jogo)
- Fix bug that caused preexisting Python loggers to be disabled when using MLflow with the SQLAlchemyStore (#3653, @arthury1n)
- Fix ``h5py`` library incompatibility for exported Keras models (#3667, @tomasatdatabricks)

> - Add ``DatabricksModelArtifactRepository`` (#3643, @arjundc-db)

Reviewer: Databricks-specific. Let’s move it into small changes

Reviewer 2: Done!

Changelog
=========
1.13 (2020-12-22)
-----------------
MLflow 1.13 includes several major features and improvements:

Features:

New fluent APIs for logging in-memory objects as artifacts:

> - Add `mlflow.log_text` (#3678, @harupy)

Reviewer 3: @harupy as mentioned in https://github.com/mlflow/mlflow/pull/3790, this might introduce an API inconsistency:

> I am not sure why the file path argument [in log_text() and friends] was named `artifact_file`, as my impression is that `path` and `artifact_path` are far more common in the source code (1400+ vs 80+ matches). I think the argument `artifact_file` in the Python API should be renamed to `artifact_path` or `path` for consistency with `mlflow.log_artifacts()` / `list_artifacts()` and the rest of the API. I think this is still possible without breaking the released API, as this change was not yet released. Might also be it was deliberately named `artifact_file`, I don't know. 

Sorry for bothering if this was intentional or not a concern.

Reviewer 2: Thanks for the commemt. `artifact_path` represents a directory, not a file path. That’s why we chose `artifact_file`.

Reviewer 3: I see, thanks for the explanation.

Reviewer 2: We also considered the following design:

```python
def log_xxx(obj, filename, artifact_path):
    ...
```

This would enforce users to specify two arguments when loggigng an artifact in a subdirectory, which seems a bit inconvenient.

- Add ``mlflow.log_dict`` which logs a dictionary as an artifact (#3685, @harupy)
- Add ``mlflow.log_figure`` which logs a figure object as an artifact (#3707, @harupy)
- Add ``mlflow.log_image`` which logs an image object as an artifact (#3728, @harupy)

> UI update / fixes (#3867, @smurching):

Reviewer: ```suggestion
UI updates / fixes (#3867, @smurching):
```

> - Add "Empty state" overlay for experiments

Reviewer: This is a small change. Let's move it down to the "Small changes, bug fixes, and doc updates" section (see https://github.com/mlflow/mlflow/pull/3890/files#r548237542 for context)

Reviewer 2: Done!

- Add ``SectionErrorBoundary`` for better error handling

> - Use local storage to persist model registry UI settings

Reviewer: ```suggestion
- Model registry UI settings are now persisted across browser sessions
```

> - Add ``data-test-id`` / DOM classname to facilitate e2e UI testing

Reviewer: Lines 19 and 21 here don't seem like noteworthy features. 19 is more of a bug fix and 21 is a dev infrastructure item that probably shouldn't be called out explicitly

Reviewer 2: Removed both of them and added #3867 (which made these changes) in `small changes`.

> - Add description to model version table

Reviewer: ```suggestion
- Add model version ``description`` field to model version table
```

- Add model version link in compact experiment table view
- Add logged/registered model links in experiment runs page view

> - Enhance artifact viewer for MLflow models

Reviewer: Can we call these out before the other UI items (i.e. move them to the top of this section)? They seem to be the most significant.

Reviewer 2: Done!

Autologging enhancements:

> - Add ``exclusive`` flag to remaining relevant autologging integrations (#3869, @dbczumar)

Reviewer: ```suggestion
- Add ``exclusive`` configuration flag to autologging integrations (#3869, @dbczumar)
```

Reviewer: @harupy We're also introducing a `disabled` flag as part of the release, but I don't see that called out in the changelog. Did we miss the PR for it?

Reviewer 2: @dbczumar I think it's because #3815 is marked as `rn/none`. I'll manually move it here.

- Set tags on autologged runs for easy identification (and add tags to start_run) (#3847, @dbczumar)
- Add ``log_models`` configuration for autologging (#3663, @mohamad-arabi)

More features and improvements:

- Retry SQLAlchemy engine creation (#3784, @HCoban)
- Add support for ``statsmodels`` flavor (#3304, @olbapjose)

> - Make ``run_id`` an optional parameter for python client (#3768, @wentinghu)

Reviewer: #3768 only applies to the `MlflowClient.create_model_version` API; I don't think this needs to be called out as a feature. I'd classify it as a small bug fix (the fact that `run_id` was required previously was a bug, IMO)

Reviewer 2: Done!

- Add support for nested-run in mlflow R client (#3765, @yitao-li)

> - Java client changes to incorporate PreSigned URL support (#3649, @arjundc-db)

Reviewer: We don't need to call this out, as it powers a Databricks-specific integration that is unlikely to be relevant for most MLflow users. Let's move it to the "Other bug fixes / docs" section

> - Add ``DatabricksModelArtifactRepository`` (#3643, @arjundc-db)

Reviewer: We don't need to call these out, as they power a Databricks-specific integration that is unlikely to be relevant for most MLflow users. Let's move them to the "Other bug fixes / docs" section.

Reviewer 2: makes sense. moved it in the `Bug fixes and documentation updates` section

> - Some additional model handling to ``azureml.deploy`` (#3419, @trangevi)

Reviewer: ```suggestion
- Deploying a model using ``mlflow.azureml.deploy`` now integrates better with the AzureML tracking/registry. (#3419, @trangevi)
```

(Suggestion adopted from the release note description in #3419)

Bug fixes and documentation updates:

> - Pin the version of mlflow in the databricks job to the same version as that of the client (#3880, @FlorisHoogenboom)

Reviewer: ```suggestion
- When running an MLflow Project on Databricks, the version of MLflow installed on the Databricks cluster will now match the version used to run the Project (#3880, @FlorisHoogenboom)
```

> - Add autologging safety utils to ``statsmodels`` & ``fastai`` (#3859, @dbczumar)

Reviewer: This is a small bug fix (the other safety utils additions aren't called out in this section either, which is fine - I think they all belong in small bugfixes)

Reviewer 2: Done!

- Fix bug where metrics are not logged for single-epoch ``tf.keras`` training sessions (#3853, @dbczumar)

> - Reject bool metric value (#3822, @HCoban)

Reviewer: ```suggestion
- Reject boolean types when logging MLflow metrics (#3822, @HCoban)
```

> - Fix bug related with bad alignment of history entries when ``initial_epoch`` is different from zero. (#3575, @garciparedes)

Reviewer: ```suggestion
- Fix alignment of Keras / ``tf.Keras`` metric history entries when ``initial_epoch`` is different from zero. (#3575, @garciparedes)
```

> - Unpin TF / Keras and fix bugs in autologging integrations for newer versions (#3735, @dbczumar)

Reviewer: ```suggestion
- Fix bugs in autologging integrations for newer versions of TensorFlow and Keras (#3735, @dbczumar)
```

- Drop global ``filterwwarnings`` module at import time (#3621, @jogo)

> - Only import ``matplotlib`` where necessary for artifact plotting (#3703, @dbczumar)

Reviewer: This is a small bug fix and can be moved down

Reviewer 2: Done!

> - Do not disable existing logger when initializing ``alembic`` (#3653, @arthury1n)

Reviewer: ```suggestion
- Fix bug that caused preexisting Python loggers to be disabled when using MLflow with the SQLAlchemyStore (#3653, @arthury1n)
```

> - Fix library incompatibility for exported Keras models (#3667, @tomasatdatabricks)

Reviewer: ```suggestion
- Fix ``h5py`` library incompatibility for exported Keras models (#3667, @tomasatdatabricks)
```

> - Allow Keras models to be saved with ``SavedModel`` format (#3552, @skylarbpayne)

Reviewer: This is a feature & should be moved up

Reviewer 2: Done!

> Small bug fixes and doc updates (#3887, #3882, #3845, #3861, #3855, #3854, #3833, #3828, #3826, #3825, #3800, #3809, #3786, #3794, #3731, #3776, #3760, #3771, #3754, #3750, #3749, #3747, #3736, #3701, #3699, #3658, #3675, @harupy; #3723, @mohamad-arabi; #3650, #3655, @shrinath-suresh; #3850, #3753, @dmatrix; #3664, @smurching; #3681, @sueann; #3619, @andrewnitu; #3837, @javierluraschi; #3721, @szczeles; #3653, @arthury1n; #3883, #3874, #3870, #3877, #3878, #3844, #3815, #3682, @dbczumar)

Reviewer: ```suggestion
Small changes, bug fixes and doc updates (#3887, #3882, #3845, #3861, #3855, #3854, #3833, #3828, #3826, #3825, #3800, #3809, #3786, #3794, #3731, #3776, #3760, #3771, #3754, #3750, #3749, #3747, #3736, #3701, #3699, #3658, #3675, @harupy; #3723, @mohamad-arabi; #3650, #3655, @shrinath-suresh; #3850, #3753, @dmatrix; #3664, @smurching; #3681, @sueann; #3619, @andrewnitu; #3837, @javierluraschi; #3721, @szczeles; #3653, @arthury1n; #3883, #3874, #3870, #3877, #3878, #3844, #3815, #3682, @dbczumar)
```
